sound/soc/sof/intel/byt.c

```diff
@@ -360,17 +360,6 @@ static irqreturn_t byt_irq_thread(int irq, void *context)
 	return IRQ_HANDLED;
 }
 
-static int byt_is_ipc_ready(struct snd_sof_dev *sdev)
-{
-	u64 ipcx;
-
-	ipcx = snd_sof_dsp_read64(sdev, BYT_DSP_BAR, SHIM_IPCX);
-	if ((ipcx & SHIM_BYT_IPCX_BUSY) || (ipcx & SHIM_BYT_IPCX_DONE))
-		return 0;
-
-	return 1;
-}
-
```

Comment:
ipcx & SHIM_BYT_IPCX_DONE is removed, so cmd_done function should be refined because SHIM_BYT_IPCX_DONE is set for is_ipc_ready. And please refine other platforms.

Reply:
This is my advice.  You could also refer to this branch https://github.com/thesofproject/linux/tree/archive/topic/sof-dev-20180619/sound/soc/sof/intel to get the original code. We added some code to fix IPC bugs, now they are not needed. Please only change the code for host sending msg to FW,  not FW to host now.

````
-       if (ipcx & SHIM_BYT_IPCX_DONE &&
-           !(imrx & SHIM_IMRX_DONE)) {
-               /* Mask Done interrupt before first */
-               snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR,
-                                                  SHIM_IMRX,
-                                                  SHIM_IMRX_DONE,
-                                                  SHIM_IMRX_DONE);
-               /*
-                * handle immediate reply from DSP core. If the msg is
-                * found, set done bit in cmd_done which is called at the
-                * end of message processing function, else set it here
-                * because the done bit can't be set in cmd_done function
-                * which is triggered by msg
-                */
-               if (snd_sof_ipc_reply(sdev, ipcx))
-                       byt_cmd_done(sdev, SOF_IPC_DSP_REPLY);
+       if (ipcx & SHIM_BYT_IPCX_DONE) {
+
+               snd_sof_ipc_reply(sdev, ipcx);
+
+                               /* clear DONE bit - tell DSP we have completed */
+               snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR, SHIM_IPCX,
+                                                  SHIM_BYT_IPCX_DONE, 0);
+
+           /* unmask Done interrupt */
+               snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR, SHIM_IMRX,
+                                                  SHIM_IMRX_DONE, 0)
        }
 
        /* new message from DSP */
@@ -427,14 +422,6 @@ static int byt_cmd_done(struct snd_sof_dev *sdev, int dir)
                /* unmask busy interrupt */
                snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR, SHIM_IMRX,
                                                   SHIM_IMRX_BUSY, 0);
-       } else {
-               /* clear DONE bit - tell DSP we have completed */
-               snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR, SHIM_IPCX,
-                                                  SHIM_BYT_IPCX_DONE, 0);
-
-               /* unmask Done interrupt */
-               snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR, SHIM_IMRX,
-                                                  SHIM_IMRX_DONE, 0);
        }
````

Reply:
Hm, I think there's a bug in my patch set. Looks like there are indeed cases when snd_sof_ipc_reply() has to return a non-zero value, specifically when no message is being transmitted. I'll fix that and will think about your comments. Thanks!

Reply:
Concerning your comment. It seems to make sense to me, but I'd postpone it to a separate patch or a patch set. With this patch set I tried to change the actual control flow as little as possible. This would be a significant change. So, unless it is really tightly coupled to this patch set, I'd rather postpone it to a later PR.